### ISSUE TITLE
Supporting DR.DECR processing pipeline in ColBERT V2

### ISSUE THREAD

**Issue by jdpsen** [primeqa/primeqa#233]
_franzm created the following on May 13:_  

 Here are the initial results on DR.DECR processing pipeline using ColBERT V2 code.

The runs use the following training data:
- NQ: `/u/franzm/ws/nq/ws/g/gd/gda/ColBERT.C3_3_20_biased200_triples_text.tsv`
- XOR-TyDi: 
`/dccstor/colbert-ir/franzm/data/xorqa/xorqa.train_ir_negs_100_poss_3_5epochs.tsv`
`/dccstor/colbert-ir/franzm/data/xorqa/xorqa.train_ir_negs_100_poss_3_enq_5epochs.tsv`
- parallel passages
`/dccstor/colbert-ir/yulongl/data/parallel_corpus/shuffled/notrim/NAACL2022/7lan_notrim_triple_2ep.other.clean`
`/dccstor/colbert-ir/yulongl/data/parallel_corpus/shuffled/notrim/NAACL2022/7lan_notrim_triple_2ep.en.clean`

Performance is tested on the XOR-TyDi dev set (`/dccstor/bsiyer5/ColBERT/xortydi/data/xortydi_dev.tsv`, `/dccstor/bsiyer5/ColBERT/xortydi/data/xorqa_dev_gmt.tsv`).

Hyper-parameter settings are as in the "domain supervision" XOR-TyDi runs in the DR.DECR paper, so in terms of adaptation, the runs should be considered somewhere between "zero-shot" and "with adaptation" because or differences between ColBERT V1 and V2 (different version of Transformers, compressed index, etc.)  

Additionally, the student starting point is based on NQ training only, as opposed to NQ->XOR used in the original DR.DECR, the difference in the performance of the starting point is small (2.4 points in [https://github.ibm.com/mnlp-exp/ColBERT/issues/89](url)).
 
| R@5kt | description |
|-------|-------|
| 44.9 | NQ 1 epoch, student starting point and teacher, tested on non-En |
| 71.5 | NQ 1 epoch, student starting point and teacher, tested on En |
| 64.0 | ST, NQ 1 epoch ST start, 5 epochs top level |
| 62.7 | ST, NQ 1 epoch ST start, 2 epochs token level |
| 68.9 | ST, NQ 1 epoch ST start, 2 epochs token level -> 5 epochs top level |

The model after distillation stages is 2.6 points behind the teacher (68.9 vs 71.5).

For comparison, the DR.DECR results using **ColBERT V1** based on in-domain hyperparameter tuning are:

| R@5kt | description |
|-------|-------|
| 76.3 | teacher tested on En |
| 47.7 | NQ -> XOR student |
| 63.6 | top level ST |
| 68.6 | token level ST |
| 73.1 | token level -> top level ST |

Here the model after distillation stages is 3.2 points behind the teacher (73.1 vs 76.3), while the teacher is 4.8 points stronger for V1 vs V2.

The causes of the differences between V1 and V2 will have to be investigated, but they are probably a combination of careful hyper-parameter tuning (domain supervision) used in V1, and index compression in V2.

Here are additional results, aiming to improve the teacher model performance:

| R@kt | LR | description |
|-------|-------|-------|
| 70.9 | 7.0e-07 | NQ 1 epoch |
| 71.5 | 1.5e-06 | NQ 1 epoch |
| 71.7 | 3.0e-06 | NQ 1 epoch |
| 69.5 | 6.0e-06 | NQ 1 epoch |
| 65.7 | 1.2e-05 | NQ 1 epoch |
| 72.5 | 1.5e-06 | NQ 1 epoch, in-batch negatives |

Here are additional results comparing the performance of the teacher model trained with in-batch negatives for various learning rates:

| R@5kt | LR | description |
|-------|-------|-------|
| 71.7 | 3.0e-06 | NQ 1 epoch, best LR |
| 71.9 | 6.0e-07 | NQ 1 epoch, in-batch negatives |
| 72.5 | 1.5e-06 | NQ 1 epoch, in-batch negatives |
| 72.5 | 3.0e-06 | NQ 1 epoch, in-batch negatives |
| 71.0 | 6.0e-06 | NQ 1 epoch, in-batch negatives |

 

_Original issue: https://github.ibm.com/ai-foundation/PrimeQA/issues/198_

**Comment by franzmpub** [comment]
Here are **additional results**, investigating:

1. Using ColBERT V2 model with V1 indexing and search
2. Level of compression in V2
3. Changing the fast match hyper-parameters in V2

**1. ColBERT V2 model with V1 indexing and search** yields R@5kt **74.6** versus **72.5** with V1 indexing and search .

**2. ColBERT V2 compression**

Changing number of bits used for centroid representation.

- **1 bit**

| R@5kt | iterations | partitions | clustering time [min] | faiss |
|-------|-------|-------|-------|-------|
| 72.48 | 20 | 262144 | 106 | 1.7.0 |

- **2 bits**

| R@5kt | iterations | partitions | clustering time [min] | faiss |
|-------|-------|-------|-------|-------|
| 73.7 | 20 | 50000 | 7 | 1.7.0 |
| 73.0 | 20 | 100000 | 11 | 1.7.0 |
| 73.7 | 20 | 262144 | 67 | 1.7.0 |

- **4 bits**

| R@5kt | iterations | partitions | clustering time [min] | faiss |
|-------|-------|-------|-------|-------|
| 73.7 | 20 | 50000 | 20 | 1.7.0 |
| 73.4 | 10 | 50000 | 4 | 1.7.0 |
| 73.4 | 4 | 100000 | 5 | 1.7.0 |
| 73.5 | 20 | 100000 | 20 | 1.7.0 |
| 74.0 | 20 | 262144 | 127 | 1.7.0 |
| 73.5 | 10 | 50000 | 845 | 1.7.2 |


**3. Changing V2 fast match hyper-parameters**  
Using 4 bits, 20 iterations, 262144 partitions

| R@5kt | nprobe | ncandidates [k] |
|-------|-------|-------|
| 74.0 | 4 | 8 |
| 73.8  | 8 | 8 |
| 74.1 | 4 | 16 |
| 74.3 | 4 | 32 |
| 74.2 | 8 | 32 |

**Comment by franzmpub** [comment]
Here are additional results, based on using the best teacher model (R@kt 74.3), 4 bits in residual representation, 
nprobe 4, ncandidates 32768 in search.
The last column shows the results for scoring using PLAID.

| V1 | V2 | V2, updated | V2, PLAID | description |
|-------|-------|-------|-------|-------|
| 76.3 | 71.5 | 74.3 | 74.5 | teacher tested on En |
| 73.1 | 69.9 | 71.1 | 71.2 | ST, final |

**Comment by yulongl-ibm** [comment]
Experiments to improve teacher in Colbert V2

V1 | NQ 1ep 1.5e-6 (Martin) | NQ 2ep 1.5e-6 | NQ 2ep 3e-6 | NQ(1ep) -> XOR (1ep) 1.5e-6 | NQ(1ep) -> XOR (1ep) 3e-6
-- | -- | -- | -- | -- | --
76.3 | 74.3 | **75.4** | 73.9 | 73.9 | 71.3

Parameters for indexing were same as Martin's best experiment. 
Experiment running now: train with NQ for even more epochs.

**Comment by yulongl-ibm** [comment]
More experiments to find the best training parameters for teacher: 
Starting from Martin's 74.3 teacher
V1 | NQ 66K -> XOR 2K | NQ 66K -> XOR 6K | NQ 66K -> XOR 8K(last week) | NQ 66K -> XOR 16K
-- | -- | -- | -- | --
76.3 | 74.8 | **75.6** | 75.4 | 75.3

Training with more NQ data and different learning rate:
NQ 120K lr8e-7 | NQ 120K lr1.5e-6 | NQ 120K lr3e-6
-- | -- | --
73.8 | 73.7 | 73.8

**Comment by franzmpub** [comment]
@yulongl-ibm : in the table above, does "66k" mean 66k NQ training steps, i.e. roughly 70% of one epoch? (and similarly for XOR, where 1 epoch is 17k steps?)

**Comment by yulongl-ibm** [comment]
Using 75.6 teacher to train student in Colbert V2
Same process as colbert V1 training 
71465 steps of parallel data 
87618 steps of XOR data

description | V1 | V2 old result | V2 new result
-- | -- | -- | --
teacher tested on En | 76.3 | 71.5 | 75.6
ST, final | 73.1 | 69.9 | **73.8**

**Result is 0.7 point higher than V1 best student score.** 

Language | Score
-- | --
te | 84.0
bn | 84.9
fi | 70.7
ja | 64.7
ko | 72.3
ru | 69.6
ar | 70.6
**Avg** | **73.8**
